src/js/form.js

```diff
@@ -544,8 +545,9 @@ Form.prototype.getRelatedNodes = function( attr, filter, updated ) {
     }
 
     // If a new repeat was created, update the cached collection of all form controls with that attribute
-    // If a repeat was deleted ( update.repeatPath && !updated.cloned), rebuild cache
-    if ( !this.all[ attr ] || ( updated.repeatPath && !updated.cloned ) ) {
+    // If a repeat was deleted ( update.repeatPath && !updated.cloned), rebuild cache.
+    // Exclude outputs from the cache, because outputs can be added via itemsets (in labels).
+    if ( !this.all[ attr ] || ( updated.repeatPath && !updated.cloned ) || filter === '.or-output' ) {
```

Comment:
More of a general question than anything particular to this change/review: it's concerning to me that responsibilities for discrete elements/widgets like this are spread out across so many parts of the codebase.

I wonder if we could gradually move toward a model where those discrete things provide an interface for these kinds of conditionals/branches and the form could delegate to those? I'm not sure how that would work with this specifically, but it's something I've been thinking about generally.

Reply:
Yes, this is concerning to me too. It starts off elegant and then slowly grows into a monster. This is especially the case with performance tweaks like this (no caching at all would create much cleaner code). Definitely could use some redesign.